Synchronize AutoCloseableCloser

pettyjamesm opened the pull request:
## Description
Synchronizes access to the internal structure of AutoCloseableCloser, which previously allowed racey access to the underlying Deque of registered entries.

This was mostly an issue in test methods that could run concurrently with one another.


<!-- Mark the appropriate option with an (x). Propose a release note if you can.
More info at https://trino.io/development/process#release-note -->
## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

dain reviewed (commented):
I'm concerned about closing while holding a lock resulting in a deadlock or such.  I'd either use a concurrent Deque, or something like this
```java
@GuardedBy("this")
private boolean closed;

    public <C extends AutoCloseable> C register(C closeable)
    {
        requireNonNull(closeable, "closeable is null");
        synchronized {
            if (!closed) {
               stack.addFirst(closeable);
               return this;
          }
          closeable.close(); // not sure if this would break stuff
          return closeable;
    }
```

Then in close you set `closed` copy the deque and close everything.  We could also switch to an array list in this case to save on memory.

pettyjamesm commented:
Updated the implementation. I don't think holding the lock would have been an issue here, but I switched it to throw an exception when registering after close has been called and to copy the deque contents out before closing. This is no longer exactly consistent with `com.google.common.io.Closer` which this class claims to be "inspired by" but I guess "inspired by" doesn't require matching the semantics exactly.

dain commented on lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/AutoCloseableCloser.java:
```diff
@@ -26,6 +30,9 @@
+    @GuardedBy("this")
+    private boolean closed;
+    @GuardedBy("this")
     private final Deque<AutoCloseable> stack = new ArrayDeque<>(4);
```
I think this can be an ArrayList now... if you reverse the list in the close method

pettyjamesm replied:
Done
